### PR TITLE
pybind/*/setup.py: fix clang detection with ccache

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Samuel Just <sjust@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
otherwise we have
```
master/build/src/pybind/rados/rados.c -o /home/jenkins-build/build/workspace/ceph-perf-crimson/ceph-master/build/lib/cython_modules/temp.linux-x86_64-3.6/home/jenkins-build/build/workspace/ceph-perf-crimson/ceph-master/build/src/pybind/rados/rados.o -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv
clang-8: error: unknown argument: '-fstack-clash-protection'
```

**Comment by tchaikov** [comment]
see https://jenkins.ceph.com/job/ceph-perf-crimson/164/console

File: src/pybind/cephfs/setup.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def get_python_flags(libs):
         libraries=libs + py_libs,
         extra_compile_args=filter_unsupported_flags(
             compiler.compiler[0],
-            distutils.sysconfig.get_config_var('CFLAGS').split()),
+            compiler.compiler[1:] + distutils.sysconfig.get_config_var('CFLAGS').split()),
```
Comment (rzarzynski) [inline src/pybind/cephfs/setup.py:75]: ACK: [On platforms with a command-line (Unix, DOS/Windows), each of these is a string that will be split into executable name and (optional) list of arguments](https://docs.python.org/2/distutils/apiref.html#distutils.ccompiler.CCompiler.set_executables).

File: src/pybind/cephfs/setup.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def get_python_flags(libs):
         libraries=libs + py_libs,
         extra_compile_args=filter_unsupported_flags(
```
Comment (rzarzynski) [inline src/pybind/cephfs/setup.py:73]: ACK: `-fstack-clash-protection` came from the [`new_compiler()`](https://docs.python.org/2/distutils/apiref.html#distutils.ccompiler.new_compiler) call.
`filter_unsupported_flags()` just squeezes some compiler options if it detects `clang`.

```python
def filter_unsupported_flags(compiler, flags):
    args = takewhile(lambda argv: not argv.startswith('-'), [compiler] + flags)
    if any('clang' in arg for arg in args):
        return list(filterfalse(lambda f:
                                f in ('-mcet',
                                      '-fstack-clash-protection',
                                      '-fno-var-tracking-assignments',
                                      '-Wno-deprecated-register',
                                      '-Wno-gnu-designator') or
                                f.startswith('-fcf-protection'),
                                flags))
    else:
        return flags
```

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-clang-cache-kefu/e2d2ad7a5f05627055d80dd6307572e5f5c0e7b0/

**Comment by tchaikov** [comment]
@athanatos sorry for stealing your commits and merging it without your blessings. i really want to test and merge https://github.com/ceph/ceph-build/pull/1576 sooner.

**Comment by athanatos** [comment]
No worries, I hadn't gotten around to making sure that was the right
solution ;).
-Sam

**Pull request merged**